### ISSUE TITLE
[QUESTION] Gradients wrt. to vertex position in wp.sim

### ISSUE THREAD

**Issue by JonathanKuelz** [NVIDIA/warp#610]
Is there a way to compute gradients with respect to the position of mesh vertices in a rigid body simulation based on the `wp.sim` module?

Issue #534 seems to indicate that it is possible for `wp.fem`. However, the `handle_contact_pairs` kernel in `wp.sim.collide` uses `wp.mesh_get`, which, according to the documentation, does not yet support gradients. Is there a workaround / any way to get these gradients?

Edit: This was also discussed but never answered in #233 .

**Comment by JonathanKuelz** [comment]
It seems that -- as long as `model.requires_grad` is True -- these gradients are available if one manually sets the `requires_grad` of the underlying `wp.mesh` to True. There is no exposed interface to do so, but it can easily be done, e.g., by implementing a wrapper around `wp.sim.Mesh`:

```python
class DiffMesh(wp.sim.Mesh):

    def finalize(self, device=None):
        """A copy-past of wp.sim.Mesh.finalize() with the single change that requires_grad=True is set for pos."""
        with wp.ScopedDevice(device):
            pos = wp.array(self.vertices, dtype=wp.vec3, requires_grad=True)
            vel = wp.zeros_like(pos, requires_grad=True)
            indices = wp.array(self.indices, dtype=wp.int32)

            self.mesh = wp.Mesh(points=pos, velocities=vel, indices=indices)
            return self.mesh.id
```

Preliminary results seem to be reasonable -- however, as `wp.mesh_get` is still used in this process (see above), I was wondering if this is a valid solution or if this might break in certain cases I have yet to experience?

**Comment by zlee-personal** [comment]
I was wondering something similar. Would the correct place to back propagate to be `model.shape_geo_src[0].mesh._points`?

**Comment by JonathanKuelz** [comment]
Short update: While the initial results I obtain seem to be somewhat promising, I am encountering bugs I cannot explain, e.g., gradients being zero when they clearly shouldn't be. I can't 100% confirm this is due to the workaround posted above, but as longs as there's not other way to backpropagate gradients to mesh vertices it is my best guess.
